src/node.rs

```diff
@@ -411,520 +412,549 @@ impl PbftNode {
+        // Commit the new block using one of the parsed messages and skip straight to Finished
+        self.service
+            .commit_block(messages[0].get_block().block_id.clone())
+            .map_err(|e| PbftError::InternalError(format!("Failed to commit block: {:?}", e)))?;
+        state.phase = PbftPhase::Finished;
 
-        Ok(message.get_info().get_signer_id().to_vec())
+        // Call on_block_commit right away so we're ready to catch up again if necessary
+        self.on_block_commit(BlockId::from(messages[0].get_block().get_block_id()), state);
+
+        Ok(())
     }
 
-    /// Verifies the consensus seal from the current block, for the previous block
-    fn verify_consensus_seal(
+    /// Handle a `BlockValid` update from the Validator
+    ///
+    /// This message arrives after `check_blocks` is called, signifying that the validator has
+    /// successfully checked a block with this `BlockId`. Once a `BlockValid` is received for the
+    /// working block, transition to the Preparing phase.
+    #[allow(clippy::ptr_arg)]
+    pub fn on_block_valid(
         &mut self,
-        block: &Block,
+        block_id: &BlockId,
         state: &mut PbftState,
     ) -> Result<(), PbftError> {
-        // We don't publish a consensus seal until block 1, so we don't verify it
-        // until block 2
-        if block.block_num < 2 {
-            return Ok(());
-        }
+        debug!("{}: <<<<<< BlockValid: {:?}", state, block_id);
+        let block = match state.working_block {
+            Some(ref block) => {
+                if &BlockId::from(block.get_block_id()) == block_id {
+                    Ok(block.clone())
+                } else {
+                    warn!("Got BlockValid that doesn't match the working block");
+                    Err(PbftError::NotReadyForMessage)
+                }
+            }
+            None => {
+                warn!("Got BlockValid with no working block");
+                Err(PbftError::NoWorkingBlock)
+            }
+        }?;
 
-        if block.payload.is_empty() {
-            return Err(PbftError::InternalError(
-                "Got empty payload for non-genesis block!".into(),
-            ));
-        }
+        state.switch_phase(PbftPhase::Preparing);
+        self._broadcast_pbft_message(
+            state.seq_num,
+            &PbftMessageType::Prepare,
+            block.clone(),
+            state,
+        )?;
+        Ok(())
+    }
 
-        let seal: PbftSeal =
-            protobuf::parse_from_bytes(&block.payload).map_err(PbftError::SerializationError)?;
+    /// Handle a `BlockCommit` update from the Validator
+    ///
+    /// A block was sucessfully committed; update state to be ready for the next block, make any
+    /// necessary view and membership changes, garbage collect the logs, update the commit & idle
+    /// timers, and start a new block if this node is the primary.
+    #[allow(clippy::needless_pass_by_value)]
```

Comment:
Can't we just pass in the reference?

Reply:
Was copied as-is, fixed in new commit: "Remove needless_pass_by_value allows"